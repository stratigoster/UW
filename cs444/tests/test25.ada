-- no errors here
-- WARNING: some crazy pointer tests ahead
PACKAGE main IS
  type data;
  type x;
  type z;
  type y is access x;
  type x is access z;
  type z is access data;
  type pdata is access data; -- forward references work!
  type data is record
    a : integer;
    b : boolean;
    p : pdata;
    blah : x;
  end record;
  hi : x;
  p : pdata;
  i : integer := hi.a;
  j : integer := p.a; -- don't need to use `all'
  l : data := p.all.p.all.p.all.p.all.p.all.p.p; -- lots of dereferencing
  k : data := hi.blah; -- hi.blah == hi.blah.all.all which is of type data
  q : pdata := p; -- no dereferencing here
  r : data := p; -- note automatic dereferencing

  type a;
  type b;
  type a is access b; -- circular types; pretty pointless, but is legal
  type b is access a;

  boo : a;
  hoo : b := boo; -- this works :)
END main;

PACKAGE BODY main IS
BEGIN
  null;
END main;
