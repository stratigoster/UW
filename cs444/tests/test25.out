

#!/usr/bin/env python
import math
def _forwardGen(seq):
	for i in seq:
		yield { 'all':i }
def _reverseGen(seq):
	i = len(seq)
	while i>0:
		i = i - 1
		yield { 'all':seq[i] }
class _dummy_(Exception):pass
class _numeric_error(Exception):pass
class _constraint_error(Exception):pass
class main:
	try:
		hi = { 'all':{'all':0} }
		p = { 'all':{'all':0} }
		i = { 'all':hi['all']['all']['all']['a'] }
		j = { 'all':p['all']['all']['a'] }
		l = { 'all':p['all']['all']['p']['all']['p']['all']['p']['all']['p']['all']['p']['all']['p']['all'] }
		k = { 'all':hi['all']['all']['all']['blah']['all']['all'] }
		q = { 'all':p['all'] }
		r = { 'all':p['all']['all'] }
		boo = { 'all':{'all':0} }
		hoo = { 'all':boo['all']['all'] }
		def __init__(self):
			try:
				pass
				pass
			except _dummy_:
				pass
	except _dummy_:
		pass
main()
